backend/app/api/routers/threads.py

```diff
@@ -0,0 +1,82 @@
+from datetime import datetime
+from http import HTTPStatus
+from typing import Optional
+
+from fastapi import APIRouter, WebSocket, WebSocketDisconnect
+
+from app.api.exceptions import EntityNotFound
+from app.api.managers import WsConnectionManager
+from app.api.repositories import SupportThreadRepository
+from app.api.schemas import (
+    Filter,
+    MessageCreate,
+    SupportThread,
+    SupportThreadCreate,
+    SupportThreadPatch,
+)
+
+router = APIRouter()
+ws_manager = WsConnectionManager()
+
+
+@router.post('/threads', response_model=SupportThread, status_code=HTTPStatus.CREATED)
+async def create_thread(thread: SupportThreadCreate) -> SupportThread:
+    thread_new = await SupportThreadRepository.create(thread)
+
+    return thread_new
+
+
+@router.get('/threads', response_model=list[SupportThread])
+async def find_all_threads(flt: Optional[Filter] = None) -> list[SupportThread]:
+    threads = await SupportThreadRepository.find_all(flt)
+
+    return threads
+
+
+@router.get('/threads/{thread_id}', response_model=SupportThread)
+async def find_thread(thread_id: int) -> SupportThread:
+    thread = await SupportThreadRepository.find_by_id(thread_id)
+
+    if thread is None:
+        raise EntityNotFound('thread')
+
+    return thread
+
+
+@router.patch('/threads/{thread_id}', response_model=SupportThread)
+async def patch_thread(thread_id: int, thread_ptc: SupportThreadPatch) -> SupportThread:
+    thread = await SupportThreadRepository.find_by_id(thread_id)
+
+    if thread is None:
+        raise EntityNotFound('thread')
+
+    thread = await SupportThreadRepository.patch(thread_ptc, thread_id)
```

Comment:
Race condition: this code does two lookups, first to check if the thread exists, second to patch the thread. It is bad because the existence of a thread may change between these lookups. What you should do instead is raise the exception from within `SupportThreadRepository.patch` (or make it return some special value if the thread is not found and then raise exception after that)

Reply:
Ops, You are right it's potential place for race condition, I will change the return type of the patch function to Optional[SupportThread], so if a thread was not found in the repository level, then return None, and it will be handled in router level

Reply:
I think they are more or less equivalent, as long as you document, how the function actually behaves if a thread is not found. Because you have your own exception class dedicated to that error specifically, I think, raising an exception from a repository function would be logical 👍